homeassistant/components/delijn/sensor.py

```diff
@@ -0,0 +1,107 @@
+"""Support for De Lijn (Flemish public transport) information."""
+import asyncio
+import logging
+
+import voluptuous as vol
+
+from homeassistant.components.sensor import PLATFORM_SCHEMA
+from homeassistant.exceptions import PlatformNotReady
+from homeassistant.helpers.aiohttp_client import async_get_clientsession
+import homeassistant.helpers.config_validation as cv
+from homeassistant.helpers.entity import Entity
+
+_LOGGER = logging.getLogger(__name__)
+
+CONF_NEXT_PASSAGE = 'nextpassage'
+CONF_STOP_ID = 'stop_id'
+CONF_API_KEY = 'api_key'
+CONF_MAX_PASSAGES = 'max_passages'
+
+DEFAULT_NAME = 'De Lijn'
+
+PLATFORM_SCHEMA = PLATFORM_SCHEMA.extend({
+    vol.Required(CONF_API_KEY): cv.string,
+    vol.Required(CONF_NEXT_PASSAGE): [{
+        vol.Required(CONF_STOP_ID): cv.string,
+        vol.Optional(CONF_MAX_PASSAGES, default=5): cv.positive_int}]
+})
+
+
+async def async_setup_platform(
+        hass, config, async_add_entities, discovery_info=None):
+    """Create the sensor."""
+    from pydelijn.api import Passages
+
+    api_key = config.get(CONF_API_KEY)
+    name = DEFAULT_NAME
+
+    session = async_get_clientsession(hass)
+
+    sensors = []
+    for nextpassage in config.get(CONF_NEXT_PASSAGE):
+        stop_id = nextpassage[CONF_STOP_ID]
+        max_passages = nextpassage[CONF_MAX_PASSAGES]
+        line = Passages(hass.loop, stop_id, max_passages, api_key, session)
+        sensors.append(DeLijnPublicTransportSensor(line, name))
+
+    tasks = [sensor.async_update() for sensor in sensors]
+    if tasks:
+        await asyncio.wait(tasks)
+    if not all(sensor._attributes for sensor in sensors):
+        raise PlatformNotReady
+
+    async_add_entities(sensors, True)
+
+
+class DeLijnPublicTransportSensor(Entity):
+    """Representation of a Ruter sensor."""
+
+    def __init__(self, line, name):
+        """Initialize the sensor."""
+        self.line = line
+        self._attributes = {}
+        self._name = name
+        self._state = None
+
+    async def async_update(self):
+        """Get the latest data from the De Lijn API."""
+        await self.line.get_passages()
+        if self.line.passages is None:
+            _LOGGER.error("No data recieved from De Lijn")
+            return
+        try:
+            attributes = {}
+            first = self.line.passages[0]
+            self._state = first['due_in_min']
```

Comment:
We don't want relative times in the state machine. Relative times will cause not needed state updates.

I suggest to calculate the absolute utc iso timestamp and return that as state. Also set the entity property `device_class` to timestamp. That will allow to show the absolute timestamp as relative time in the frontend.

https://developers.home-assistant.io/docs/en/entity_sensor.html#available-device-classes

Reply:
Just to be sure I understand correctly. You're proposing to use a timestamp as the self._state, correct? Because the issue of the API is that their datetimes are naive. So I'd have to do some magic to make them into a UTC timestamp. Would this kind of code be ok?

`
local_timezone = pytz.timezone("Europe/Brussels")
if first['due_at_rt'] is not None:
    first_passage_naive = datetime.strptime(first['due_at_rt'], '%Y-%m-%dT%H:%M:%S')
else:
    first_passage_naive = datetime.strptime(first['due_at_sch'], '%Y-%m-%dT%H:%M:%S')
first_passage_local = local_timezone.localize(first_passage_naive)
first_passage_utc = first_passage_local.astimezone(pytz.utc)
self._state = first_passage_utc
`

The localized timestamp would be something like this: 2019-07-24 13:08:31+02:00
The UTC timestamp, that would be used for the self._state, would be something like: 2019-07-24 11:08:31+00:00

Is that what you expect to be used? Because event wit the device_class property set it doesn't seem to do much interpretation in the frontend in my dev environment.

Thanks in advance for the feedback.

Reply:
The state timestamp needs to be an iso datetime string.

Also, please use our datetime utils as much as possible.
https://github.com/home-assistant/home-assistant/blob/dev/homeassistant/util/dt.py

Reply:
Here are the lovelace config options for timestamp entity for the entity card:
https://www.home-assistant.io/lovelace/entities/#format

Reply:
Changes applied in last commit. Made use of the dtutils within home-assistant as well.